client/app/bundles/course/users/components/misc/PersonalTimeEditor.tsx

```diff
@@ -0,0 +1,290 @@
+  bonusEndAt: yup.date().nullable(),
+});
+
+const PersonalTimeEditor: FC<Props> = (props) => {
+  const { item, intl } = props;
+  const initialValues = {
+    fixed: item.fixed,
+    startAt: item.personalStartAt,
+    bonusEndAt: item.personalBonusEndAt,
+    endAt: item.personalEndAt,
+  };
+  const { control, handleSubmit, setError, reset } = useForm<IFormInputs>({
+    defaultValues: initialValues,
+    resolver: yupResolver(validationSchema),
+  });
+  const { userId } = useParams();
+  const [isCreating, setIsCreating] = useState(false);
+  const [isSaving, setIsSaving] = useState(false);
+  const [isDeleting, setIsDeleting] = useState(false);
+  const dispatch = useDispatch<AppDispatch>();
+
+  const handleCreate = (): Promise<void> => {
+    const data = {
+      id: item.id,
+      fixed: item.fixed,
+      startAt: item.itemStartAt,
+      bonusEndAt: item.itemBonusEndAt,
+      endAt: item.itemEndAt,
+    };
+    setIsCreating(true);
+    return dispatch(updatePersonalTime(data, +userId!))
+      .then(() => {
+        reset(initialValues);
+      })
+      .finally(() => {
+        setIsCreating(false);
+        toast.success(intl.formatMessage(translations.createSuccess));
+      })
+      .catch((error) => {
+        toast.error(intl.formatMessage(translations.createFailure));
+        throw error;
+      });
+  };
+
+  const onDelete = (): Promise<void> => {
+    setIsDeleting(true);
+    return dispatch(deletePersonalTime(item.personalTimeId!, +userId!))
+      .then(() => {
+        reset(initialValues);
+        toast.success(intl.formatMessage(translations.deleteSuccess));
+      })
+      .finally(() => {
+        setIsDeleting(false);
+      })
+      .catch((error) => {
+        toast.error(intl.formatMessage(translations.deleteFailure));
+        throw error;
+      });
+  };
+
+  const onSubmit = (
+    formData: IFormInputs,
+    // eslint-disable-next-line @typescript-eslint/no-shadow
+    setError: UseFormSetError<IFormInputs>,
+  ): Promise<void> => {
+    const data = {
+      ...formData,
+      id: item.id,
+    };
+    setIsSaving(true);
+    return dispatch(updatePersonalTime(data, +userId!))
+      .then(() => {})
+      .finally(() => {
+        setIsSaving(false);
+        toast.success(
+          intl.formatMessage(translations.updateSuccess, {
+            title: item.title,
+          }),
+        );
+      })
+      .catch((error) => {
+        toast.error(
+          intl.formatMessage(translations.updateFailure, {
+            error: error.response.data.errors,
+          }),
+        );
+        setReactHookFormError(setError, error.response.data.errors);
+        throw error;
+      });
+  };
+
+  return (
+    <>
+      {item.new ? (
+        <TableCell colSpan={4}>
+          <Grid container flexDirection="column" alignItems="center">
+            <LoadingButton
+              loading={isCreating}
+              onClick={handleCreate}
```

Comment:
Is it? In my local testing, adding personal time without saving will create a new entry in the `course_personal_times` table. User only needs to save if they want to edit. The default values are generated from reference timeline. Or am I looking in the wrong place?

Reply:
See this

https://user-images.githubusercontent.com/42570513/176154577-2264d943-c410-4399-880a-09c174415844.MOV

Reply:
Oh, I realize what you mean. So do you want it to revert back to the original behaviour?

The reason I implemented this is because if the user clicks on "Add personal time", it should be quite intentional that they want to add a new personal time already, so it's immediately saved. However, there's a case to be made for duplicate hits to the backend when we edit and save afterwards.

Is there a potential workflow where: 
1. User creates personal time but doesn't change anything (i.e. personal time will be created based off reference timeline)
2. User recomputes personal times for that user?

If such a workflow exists, then it would require a page refresh in the original behaviour, but with this change it can be done in a single page.